test/jdk/java/security/Security/SynchronizedAccess.java

```diff
@@ -42,52 +45,96 @@ public static void main(String[] args) throws Exception {
     }
 
     public static void main0(String[] args) throws Exception {
+        var providersCountBefore = Security.getProviders().length;
         AccessorThread[] acc = new AccessorThread[200];
-        for (int i=0; i < acc.length; i++)
-            acc[i] = new AccessorThread("thread"+i);
-        for (int i=0; i < acc.length; i++)
+        for (int i = 0; i < acc.length; i++)
+            acc[i] = new AccessorThread("thread" + i);
+        for (int i = 0; i < acc.length; i++)
             acc[i].start();
-        for (int i=0; i < acc.length; i++)
+        for (int i = 0; i < acc.length; i++)
             acc[i].join();
+
+        var providersCountAfter = Security.getProviders().length;
+        Asserts.assertEquals(providersCountBefore + 1, providersCountAfter);
     }
-}
 
-class AccessorThread extends Thread {
+    public static class AccessorThread extends Thread {
 
-    public AccessorThread(String str) {
-        super(str);
-    }
+        public AccessorThread(String str) {
+            super(str);
+        }
 
-    public void run() {
-        Provider[] provs = new Provider[10];
-        for (int i=0; i < provs.length; i++)
-            provs[i] = new MyProvider("name"+i, "1", "test");
+        public void run() {
+            Provider[] provs = new Provider[10];
+            for (int i = 0; i < provs.length; i++)
+                provs[i] = new MyProvider("name" + i, "1", "test");
 
-        int rounds = 20;
-        while (rounds-- > 0) {
-            try {
-                for (int i=0; i<provs.length; i++) {
-                    Security.addProvider(provs[i]);
+            int rounds = 20;
+            while (rounds-- > 0) {
+                try {
+                    for (int i = 0; i < provs.length; i++) {
+                        Security.addProvider(provs[i]);
+                    }
+                    Signature.getInstance("sigalg");
+                    // skipping first provider so there is always 1 available for getInstance
+                    for (int i = 1; i < provs.length; i++) {
+                        Security.removeProvider("name" + i);
+                    }
+                } catch (NoSuchAlgorithmException nsae) {
+                    throw new RuntimeException("Should not reach here " + nsae);
```

Comment:
The message could be a little clearer. Something like, "Expected provider for sigalg not found."

I would include `nsae` as a second, 'cause' parameter to the RuntimeException. That way the full stack is shown in the logs. It isn't strictly necessary here since it's pretty obvious where it's coming from, but it's a good habit to be in.